\documentclass[10pt,twocolumn]{article}
\usepackage[margin=0.6in]{geometry}
\usepackage[utf8]{inputenc}
\usepackage{authblk}
\usepackage{doi}
\usepackage{tcolorbox}
\usepackage{enumitem}
\usepackage{graphicx,pdflscape,multirow}
\usepackage{array}
\usepackage{xcolor}
\usepackage{multicol}
\usepackage{wrapfig,lipsum,booktabs}
\usepackage{fancybox}
\usepackage{amsmath}


\usepackage{hyperref}
\definecolor{darkblue}{rgb}{0.0,0.0,0.75}
\hypersetup{colorlinks,breaklinks,
            linkcolor=darkblue,urlcolor=darkblue,
            anchorcolor=darkblue,citecolor=darkblue}


\newcommand{\fixme}[1]{{\color{red} (#1)}}
\newcommand{\coloc}{\texttt{escheR}}

% \renewcommand\Authfont{\fontsize{8}{14.4}\selectfont} % change author fontsize
% \renewcommand\Affilfont{\fontsize{6}{10.8}\itshape}   % change auth affil fontsize
\makeatletter % make affiliations on one line
\renewcommand\AB@affilsepx{, \protect\Affilfont}
\makeatother

\usepackage[sort&compress,square,numbers]{natbib}
\bibliographystyle{unsrtnat}

%\setmainfont{Helvetica}
\title{\texttt{escheR}: Unified multi-dimensional visualizations with Gestalt principles}
\author[1]{Boyi Guo}
\author[1,2]{Stephanie C. Hicks\thanks{$^*$Correspondence to \url{shicks19@jhu.edu}}}
\affil[1]{Department of Biostatistics, Johns Hopkins Bloomberg School of Public Health, MD, USA}
\affil[2]{Malone Center for Engineering in Healthcare, Johns Hopkins University, MD, USA}
\date{\today}

\begin{document}
\maketitle

\vspace{-.6in}

%% ABSTRACT ============================================================================================
\section*{Abstract}
The creation of effective visualizations is a fundamental component of data analysis. Nevertheless, current methods and
tools are inadequate to visualize the novel spatially-resolved data, limiting the ability to derive new insights and discoveries. Specifically, there are no multi-dimensional \textit{in-situ} visualization that simultaneously displays multiple variables in a 2D spatial map. To address this problem, we introduce gestalt principles to the design of spatial maps, layering aesthetics to display multiple variables. The proposed visualization can be broadly applied to spatially resolved data and 2D embedding methods. We provide an open source R package \texttt{escheR}, which seamlessly integrates into the state-of-the-art spatial omics toolboxes.

\subsection*{Availability and implementation}
The open source R package \texttt{escheR} is freely available on GitHub and is being submitted to Bioconductor (\href{https://github.com/boyiguo1/escheR}{https://github.com/boyiguo1/escheR}).
% and Zenodo (\fixme{add link}).


%% INTRODUCTION ============================================================================================
\section{Introduction}
Visualization is an indispensable component of data analysis, providing clarity that connects quantitative evidence to key conclusions \cite{dagostinomcgowan_2022}. In biomedical research, visualization receives growing recognition as essential: many scientists rely on visualization to complete their cognitive process from analysis to insight, including analytic validation of automated pipelines and scientific communication \cite{odonoghue_2021}. However, an important challenge in biomedical research is how to visualize increasingly complex, multi-dimensional data \cite{odonoghue_2010}. 

Here, we focus on two types of visualizations in biomedical research, but note that the proposed ideas could be extended beyond these applications: (i) embedding visualizations, which project data into some low-dimensional embedding or mathematical space (e.g. Principal Components Analysis (PCA) \cite{hotelling_1933}, $t$-distributed Stochastic Neighbor Embedding ($t$-SNE) \cite{vandermaaten_2008}, or Uniform Manifold Approximation and Projection (UMAP) \cite{becht_2019}) and (ii) \textit{in situ} visualizations \cite{dries_2021, Lewis_2021, odonoghue_2021}, which aim to visualize molecules captured from \textit{in situ} imaging or sequencing technologies where \textit{in situ} refers to `in its original place'. Both of these visualizations represent data in a 2D space and are motivated by recent advances in experimental technologies that profile molecules, including DNA, RNA, and proteins, at a single-cell or spatial resolution \cite{kashima_2020, moffitt_2022}. These technologies are commonly referred to as single-cell/single-nucleus RNA-sequencing (sc/snRNA-seq) \fixme{add citations} and \textit{in situ} spatially-resolved transcriptomics (SRT) \fixme{add citations}.  

A common and fundamental challenge with both of these visualizations is how to visualize multi-dimensional information in a 2D space. For example, in \textit{in situ} visualizations, we often want to create a spatial map to visualize a continuous (e.g. gene expression) or discrete (e.g. cell type or spatial domain) variable representing molecular information in the original spatial location. However, it is challenging to simultaneously visualize multi-dimensional data, such as information from disparate data domains (such as expression domain and spatial domain) or disparate data modalities (such as transcriptomics and proteomics) in the same plot. Currently, best practices for this include making two different plots displayed side-by-side (\fixme{\textbf{Figure \textbf{\ref{fig:visual}A-B}}}), one for gene expression and one for spatial domains. This creates cognitive gaps on how to associate the disparate information or how to interpret the biological findings of this multi-dimensional information regarding their (micro-)environment or colocalization. While interactive visualizations \cite{keller_2021, pardo_2022, sriworarat_2023} have the potential to mitigate this challenge, they are infeasible for scientific communications in static media, such as printed work. Developing a static and unified visualization that enables the simultaneous display of multiple dimensions of information is crucial for biomedical research.

\begin{figure*}[!t]
\begin{center}
\includegraphics[width=\textwidth]{Manuscript/figure/insitu.jpg}
\caption{\small \textbf{\texttt{escheR} improves spatial visualizations following the Gestalt principles}. (\textbf{A-B}) The traditional visualization displays the colocalization plot (\textbf{A}) and the cortex layers (spatial domains) plot (\textbf{B}) side-by-side, creating challenges to connecting colocalization status to spatial domains. (\textbf{C}) The watercolor effect enables displaying spatial domains by color-coding only outlines of circles. (\textbf{D}) \texttt{escheR} enables the multi-dimensional \texttt{in-situ} visualization that simultaneously displays the cortex layers and the colocalization status, substantially improving interpretability.}
\label{fig:visual} 
\end{center}
\end{figure*}


\section{Results}

To address these challenges, here we leverage the Gestalt (German for “unified whole”) principles for design \cite{todorovic_2008, palmer_1999} as a way to visualize multi-dimensional data in 2D visualizations. We focus on the two types of data visualizations previously introduced that are widely used in biomedical research: (i) embedding visualizations and (ii) \textit{in situ} visualizations. We provide an R package, \texttt{escheR}, implementing these ideas, which is built on the state-of-the-art data visualization framework \texttt{ggplot2} in the R programming language. Finally, we comment on how these ideas could be extended to other types of visualization in biomedical research. 


\subsection{Multi-dimensional 2D visualizations with \texttt{ggplot2} and Gestalt principles}

Gestalt principles\cite{todorovic_2008, palmer_1999} refer to a set of rules describing how humans perceive and interpret visual information and are commonly applied in art and designs. Developed in the 1920s by German psychologists Max Wertheimer, Kurt Koffka and Wolfgang Kohler, these principles help humans perceive a set of individual elements as a whole. 

Here, we leverage the principles to be able to visualize multi-dimensional data in a unified 2D plot. Our approach is to use the state-of-art data visualization framework \texttt{ggplot2} \cite{ggplot2} following the grammar of graphics \cite{wilkinson_2012} and map individual variables to different aesthetics to simultaneously display disparate variables. Specifically, we apply the figure-ground segmentation \cite{peterson_2010} in displaying two variables: one variable (e.g. expression) can be plotted as color-filled circles, serving as the \textit{figure}; one variable (e.g. spatial domains) can be plotted as the backgrounds of the circles, creating a \textit{ground} for the figure. In practice, we use the combination of \texttt{color} and \texttt{fill="transparent"} to create the background layer and \texttt{fill} to create the figure layer. When necessary to display an additional layer for a third variable, \texttt{shape} can be used to add symbols such as cross (+) and asterisk (*) to highlight in the spatial map.


For adjacent circles with limited space between them to display the background color, we use an economic implementation, colored outlines for these circles (\fixme{\textbf{Figure~\ref{fig:visual}C}}), inspired by watercolor effect \cite{pinna_1987, pinna_2001}. Watercolor effect describes the phenomenon in visual perception that surface color arises from thin boundaries and hence is applied here to perceive the background color in tight space. Overall, the figure-ground segmentation creates two isolated layers in visual perception to display the two variables while maintaining the relative spatial relationship serving as a reference between the two. In addition, other fundamental principles \cite{todorovic_2008}, such as proximity, similarity, continuity, and closure, incentivize the brain to group elements and dimensions in the visualization, guaranteeing an integrative perception of the complex multi-dimensional spatial map.

Here, we provide an open-source package called \texttt{escheR} (named after the graphic artist M.C. Escher) in the R programming language \cite{R}, leading to a simplified interface to navigate the implementation of the multi-dimensional visualization in 2D space. By adapting \texttt{ggplot2} standard, \texttt{escheR} can be seamlessly integrated into many popular spatial resolved toolboxes, such as \texttt{SpatialLIBD}\cite{pardo_2022}, \texttt{Seurat}\cite{hao_2021}, \texttt{Giotto}\cite{dries_2021} to name a few, and allow further theme customization with ease.

Next, we give two use cases to exemplify some utility of the proposed spatial visualization: (i) the spatially differential gene colocalization in the human dorsolateral prefrontal cortex using spatial transcriptomics data \cite{huukimyers_2023}; (ii) multi-dimensional UMAP highlighting differential gene expression in data-driven cell clusters \cite{freytag_2020}. 



\subsection{Multi-dimensional \textit{in situ} visualization}
In a recent study investigating the molecular organization of human dorsolateral prefrontal cortex \cite{huukimyers_2023}, two schizophrenia risk genes, membrane-bound ligand ephrin A5 (\textit{EFNA5}) and ephrin type-A receptor 5 (\textit{EPHA5}), were identified to colocalize via the cell-cell communication analysis. In addition, data suggested Layer 6 was the most highly co-localized layer compared to other cortex layers. To visually examine the inference, we applied \texttt{escheR} to create a multi-dimensional \textit{in situ} spatial map that simultaneously exhibits the cortex layers (displayed with color-coded spot outlines) and the categorized colocalization status of genes \textit{FYN} and \textit{EFNA5} (displayed with color-coded spot fill). Compared to the traditional visualization where the cortex layers and the colocalization status are visualized in two side-by-side figures (\textbf{Figure \ref{fig:visual}A-B}), our proposed visualization (\textbf{Figure \ref{fig:visual}D}) enables directly mapping colocalization status to the spatial domain, simplifying the perception of two sources of information and allowing cognitive comparison across cortex layers. 

\subsection{Multi-dimensional embedding visualizations}

The application of the proposed framework is not limited to \textit{in situ} visualizations of spatially-resolved transcriptomics data. It is broadly applicable to data mapped to any 2-dimensional coordinate system to simultaneously display multiple variables. Such systems include euclidean space (including spatial coordinate as a special case) and data-driven embedding space, for example, UMAP and $t$-SNE. To demonstrate, we applied the proposed visualization to address the challenge of simultaneously displaying cluster membership and gene expression in a single-cell UMAP plot. To address the overplotting problem, previous work proposed to apply hexagonal binning strategy to display the gene expression \cite{freytag_2020}. Here, the color-coded convex hulls are used to annotate different clusters of cells (\textbf{Figure~\ref{fig:embedding}A}). However, the convex hulls create substantial overlapping areas, creating confusion when interpreting cluster memberships of hexagons in the overlapping areas. To improve the interpretability of the visualization, we replace the convex hulls with color-coded hexagons boundaries  (\textbf{Figure~\ref{fig:embedding}B}) to avoid possible membership confusion. We note that our contribution to improve the visualization is easily implemented without any modification of \texttt{schex} as both are built upon the Grammar of Graphics \cite{wilkinson_2012} standard.


\section{Discussion}
In summary, we propose an innovative multi-dimensional spatial visualization to simultaneously display multiple variables in a 2D coordinate system. Specifically, our design leverages the gestalt principle from visual perception to introduce other dimensions of a spatial map by iterative layering aesthetics. Developed upon \texttt{ggplot2}, we provide an open-source R package \texttt{escheR} that is seamlessly compatible with popular spatial resolved data analysis toolboxes. The proposed visualization has broad applications in visualizing both spot-based and image-based spatial resolved data, spatial multi-omics as well as other spatial-based data. The proposed visualization innovation highlights the importance and potential of translating theories in vision science to data science, addressing long-standing challenges in bioinformatics data visualization.


% While \texttt{escheR} is to design to display colocalization in their spatial context, the underlying principle can be broadly applied to address visualizaiton challenges in informaiton rich visualizations. The contrast color optical illusion provides allows to add another dimension of information in any spatial visualization, which can be further used to display multi-omics presence, greatly improve the interpretability. 

% our programming provides flexibility by following the grammar of graphs, especially ggplot2




%% EPILOGUE ================================================================
\section*{Abbreviations}

\begin{itemize}[nosep]
    \item \textbf{PCA}: principal component analysis
    \item \textbf{$t$-SNE}: $t$-distributed Stochastic Neighbor Embedding
    \item \textbf{UMAP}: Uniform Manifold Approximation and Projection
    \item \textbf{EFNA5}: membrane-bound ligand ephrin A5
    \item \textbf{EPHA5}: ephrin type-A receptor 5
\end{itemize}

\section*{Author contributions}

\begin{itemize}[nosep]
    \item \textbf{Boyi Guo}:  Conceptualization, Methodology, Software, Validation, Formal analysis, Investigation, Data Curation, Writing, Visualization
    \item \textbf{Stephanie C. Hicks}: Conceptualization, Resources, Writing - Review \& Editing, Visualization, Supervision, Project administration, Funding acquisition
\end{itemize}

\section*{Declarations}

\subsection*{Ethics approval and consent to participate}
Not applicable.

\subsection*{Competing interests}
The authors declare that they have no competing interests.

\subsection*{Availability of data and materials}
The spatial transcriptomics dataset was obtained from \texttt{spatialLIBD} (\href{http://research.libd.org/spatialLIBD}{research.libd.org/spatialLIBD}). The UMAP example follows the `using\_schex' vignette in the \texttt{schex} package (\href{https://www.bioconductor.org/packages/schex}{bioconductor.org/packages/schex}). The code that generates these figures is deposited at \href{https://github.com/boyiguo1/Manuscript_escheR}{github.com/boyiguo1/Manuscript\_escheR} (\textit{v1.0}).


\subsection*{Funding}
This project was supported by the National Institute of Mental Health R01MH126393 (BG, SCH), and CZF2019-002443 (SCH) from the Chan Zuckerberg Initiative DAF, an advised fund of Silicon Valley Community Foundation. All funding bodies had no role in the design of the study and collection, analysis, and interpretation of data and in writing the manuscript.

\subsection*{Acknowledgements}
BG would like to acknowledge Dr. Leonardo Collado-Torres, Louise Huuki-Myers, Dr. Lukas M. Weber, and Leon Di Stefano for their helpful comments, feedback and suggestions on \texttt{escheR} functionality.



\subsection*{Author’s information}

\begin{itemize}[nosep]
    \item Boyi Guo (ORCiD: \href{https://orcid.org/0000-0003-2950-2349}{0000-0003-2950-2349})
    \item Stephanie C. Hicks (ORCiD: \href{https://orcid.org/0000-0002-7858-0231}{0000-0002-7858-0231})
\end{itemize} 

\subsection*{Conflict of Interest} 
None declared.


%% BIBLIOGRAPHY ==========================================================

\clearpage 
% \printbibliography
\bibliography{refs}


\clearpage
\onecolumn

{\huge Supplementary Materials}

\hrule

\vspace*{0.5cm}

\begin{center}

{\Large \texttt{escheR}: Unified multi-dimensional visualizations with Gestalt principles}

\vspace*{0.75cm}

{\large Boyi Guo, Stephanie C.\ Hicks$^*$}

\vspace*{0.3cm}

{\small $^*$Correspondence to \url{shicks19@jhu.edu}}

\end{center}

\renewcommand{\figurename}{Supplementary Figure}
\renewcommand{\tablename}{Supplementary Table}
\setcounter{figure}{0}
\setcounter{table}{0}
\setcounter{section}{0}
\setcounter{page}{1}
\makeatletter
\renewcommand{\thefigure}{S\@arabic\c@figure}
\renewcommand{\thetable}{S\@arabic\c@table}
\renewcommand{\thesection}{Supplemental Note S\@arabic\c@section}
\makeatother

\vspace*{1cm}

{\bf \large Contents}

\begin{enumerate}
    \item \textbf{Supplemental Figures~S1.}
\end{enumerate}

\clearpage 

\noindent {\LARGE Supplemental Figures}


\begin{figure*}[!h]
\begin{center}
\includegraphics[width=\textwidth]{Manuscript/figure/embedding.jpg}
\caption{\small \textbf{\texttt{escheR} improves multi-dimensional embedding visualizations}. UMAP representation along $x$ and $y$ axis of \fixme{what data? just generally need to describe and cite it, if appli}. (\textbf{A}) The \texttt{schex} R/Bioconductor package uses color-coded convex hulls to annotate cell memberships, creating confusion when interpreting hexagons in overlapping hulls. (\textbf{B}) Our package \texttt{escheR} plots hexagon-specific membership to improve interpretability.}
\label{fig:embedding} 
\end{center}
\end{figure*}


\end{document}


