\documentclass[10pt,twocolumn]{article}
\usepackage[margin=0.6in]{geometry}
\usepackage[utf8]{inputenc}
\usepackage{authblk}
\usepackage{doi}
\usepackage{tcolorbox}
\usepackage{enumitem}
\usepackage{graphicx,pdflscape,multirow}
\usepackage{array}
\usepackage{xcolor}
\usepackage{multicol}
\usepackage{wrapfig,lipsum,booktabs}
\usepackage{fancybox}
\usepackage{amsmath}


\usepackage{hyperref}
\definecolor{darkblue}{rgb}{0.0,0.0,0.75}
\hypersetup{colorlinks,breaklinks,
            linkcolor=darkblue,urlcolor=darkblue,
            anchorcolor=darkblue,citecolor=darkblue}


\newcommand{\rr}[1]{{\color{red} #1}}
\newcommand{\coloc}{\texttt{escheR}}

% \renewcommand\Authfont{\fontsize{8}{14.4}\selectfont} % change author fontsize
% \renewcommand\Affilfont{\fontsize{6}{10.8}\itshape}   % change auth affil fontsize
\makeatletter % make affiliations on one line
\renewcommand\AB@affilsepx{, \protect\Affilfont}
\makeatother


\usepackage[sort&compress,square,numbers]{natbib}
\bibliographystyle{abbrv}

\setlength{\bibsep}{0pt plus 0.3ex}

\renewcommand\Affilfont{\fontsize{9}{10.8}\itshape}   % change auth affil 

%\setmainfont{Helvetica}
\title{\texttt{escheR}: Unified multi-dimensional visualizations with Gestalt principles}
\author[1]{Boyi Guo}
\author[2]{Louise A. Huuki-Myers}
\author[3,4]{Melissa Grant-Peters}
\author[2]{Leonardo Collado-Torres}
\author[1,5,6,7]{Stephanie C. Hicks\thanks{Correspondence to \url{shicks19@jhu.edu}}}
\affil[1]{Department of Biostatistics, Johns Hopkins Bloomberg School of Public Health, MD, USA} 
\affil[2]{Lieber Institute for Brain Development, Johns Hopkins Medical Campus, Baltimore, MD, USA}
\affil[3]{Genetics and Genomic Medicine, Great Ormond Street Institute of Child Health, University College London, London, UK}
\affil[4]{Aligning Science Across Parkinson’s (ASAP) Collaborative Research Network, Chevy Chase, MD, USA}
\affil[5]{Department of Biomedical Engineering, Johns Hopkins School of Medicine, Baltimore, MD, USA}
\affil[6]{Center for Computational Biology, Johns Hopkins University, Baltimore, MD, USA}
\affil[7]{Malone Center for Engineering in Healthcare, Johns Hopkins University, MD, USA}

\date{\today}

\begin{document}
\maketitle

\vspace{-.6in}

%% ABSTRACT ============================================================================================
\section*{Abstract}
The creation of effective visualizations is a fundamental component of data analysis. In biomedical research, new challenges are emerging to visualize multi-dimensional data in a 2D space, but current data visualization tools have limited capabilities. To address this problem, we leverage Gestalt principles to improve the design and interpretability of multi-dimensional data in 2D data visualizations, layering aesthetics to display multiple variables. The proposed visualization can be applied to spatially-resolved transcriptomics data, but also broadly to data visualized in 2D space, such as embedding visualizations. We provide an open source R package \texttt{escheR}, which is built off of the state-of-the-art \texttt{ggplot2} visualization framework and can be seamlessly integrated into genomics toolboxes and workflows.

\subsection*{Availability and implementation}
The open source R package \texttt{escheR} is freely available on Bioconductor (\href{https://bioconductor.org/packages/escheR}{bioconductor.org/packages/escheR}).
% and Zenodo (\fixme{add link}).


%% INTRODUCTION ============================================================================================
\section{Introduction}
Visualization is an indispensable component of data analysis, providing clarity that connects quantitative evidence to key conclusions \cite{dagostinomcgowan_2022}. In biomedical research, visualization receives growing recognition as essential: many scientists rely on visualization to complete their cognitive process from analysis to insight, including analytic validation of automated pipelines and scientific communication \cite{odonoghue_2021}. However, an important challenge in biomedical research is how to visualize increasingly complex, multi-dimensional data \cite{odonoghue_2010}. 

Here, we focus on two types of visualizations in biomedical research, but note that the proposed ideas could be extended beyond these applications: (i) embedding visualizations, which project data into some low-dimensional embedding or mathematical space (e.g. Principal Components Analysis %(PCA)
\cite{hotelling_1933}, $t$-distributed Stochastic Neighbor Embedding ($t$-SNE) \cite{vandermaaten_2008}, or Uniform Manifold Approximation and Projection (UMAP) \cite{becht_2019}) and (ii) \textit{in situ} visualizations \cite{dries_2021, Lewis_2021, odonoghue_2021}, which aim to visualize molecules captured from \textit{in situ} imaging or sequencing technologies where \textit{in situ} refers to `in its original place'. Both of these visualizations represent data in a 2D space and are motivated by recent advances in experimental technologies that profile molecules, including DNA, RNA, and proteins, at a single-cell or spatial resolution \cite{kashima_2020, moffitt_2022}. Some most popular technologies include single-cell/nucleus RNA-sequencing (sc/snRNA-seq)
\cite{aldridge_2020} and \textit{in situ} spatially-resolved transcriptomics \cite{larsson_2021}. 
% 
% \vspace{-0.15in}

A common and fundamental challenge with both of these visualizations is how to visualize multi-dimensional information in a 2D space. For example, in \textit{in situ} visualizations, we often want to create a spatial map to visualize a continuous (e.g. gene expression) or discrete (e.g. cell type or spatial domain) variable representing molecular information in the original spatial location. However, it is challenging to simultaneously visualize multi-dimensional data, such as information from disparate data domains (such as expression domain and spatial domain) or disparate data modalities (such as transcriptomics and proteomics) in the same plot. Currently, best practices for this include making two different plots displayed side-by-side (\textbf{Figure \textbf{\ref{fig:visual}A-B}}), one for gene expression and one for spatial domains. This creates cognitive gaps on how to associate the disparate information or how to interpret the biological findings of this multi-dimensional information regarding their microenvironment or colocalization. While interactive visualizations \cite{keller_2021, pardo_2022, sriworarat_2023} have the potential to mitigate this challenge, they are infeasible for scientific communications in static media, such as printed work. Developing a static and unified visualization that enables the simultaneous display of multiple dimensions of information is crucial for biomedical research.

\begin{figure*}[!t]
\begin{center}
\includegraphics[width=\textwidth]{Manuscript/figure/insitu.jpg}
\caption{\small \textbf{\texttt{escheR} enables multi-dimensional spatial visualizations following the Gestalt principles of design}. (\textbf{A-B}) The traditional visualization displays the colocalization plot of the expression of two genes \textit{EFNA5} and \textit{EPHA5} (\textbf{A}) and the spatial domains from the dorsolateral prefrontal cortex in postmorteum human brain \cite{huukimyers_2023} (\textbf{B}) side-by-side, creating challenges to cognitively connecting colocalization status to spatial domains. (\textbf{C}) The watercolor effect enables displaying spatial domains by color-coding only outlines of circles. (\textbf{D}) \texttt{escheR} enables the multi-dimensional \textit{in situ} visualization that simultaneously displays the cortex layers and the colocalization status, substantially improving interpretability.}
\label{fig:visual} 
\end{center}
\vspace{-0.3in}
\end{figure*}

% \vspace{-0.2in}

\section{Results}

To address these challenges, here we leverage the Gestalt (German for “unified whole”) principles \cite{todorovic_2008, palmer_1999} as a path to visualize multi-dimensional data in 2D visualizations. We focus on the two types of data visualizations previously introduced that are widely used in biomedical research: (i) embedding visualizations and (ii) \textit{in situ} visualizations. We provide an R package, \texttt{escheR}, implementing these ideas, which is built on the state-of-the-art data visualization framework \texttt{ggplot2} in the R programming language. Finally, we comment on how these ideas could be extended to other types of visualization in biomedical research. 


\subsection{Multi-dimensional 2D visualizations with \texttt{ggplot2} and Gestalt principles}

Gestalt principles \cite{todorovic_2008, palmer_1999} refer to a set of rules describing how humans perceive and interpret visual information and are commonly applied in art and designs. Developed in the 1920s by German psychologists Max Wertheimer, Kurt Koffka and Wolfgang Kohler, these principles help humans perceive a set of individual elements as a whole. 

Here, we leverage the principles to be able to visualize multi-dimensional data in a unified 2D plot. Our approach is to use the state-of-art data visualization framework \texttt{ggplot2} \cite{ggplot2} following the Grammar of Graphics \cite{wilkinson_2012} and map individual variables to different aesthetics to simultaneously display disparate variables. Specifically, we apply the figure-ground segmentation \cite{peterson_2010} in displaying two variables: one variable (e.g. expression) can be plotted as color-filled circles, serving as the \textit{figure}; one variable (e.g. spatial domains) can be plotted as the backgrounds of the circles, creating a \textit{ground} for the figure. In practice, we use the combination of \texttt{color} and \texttt{fill="transparent"} to create the background layer and \texttt{fill} to create the figure layer. When necessary to display an additional layer for a third variable, \texttt{shape} can be used to add symbols such as cross (+) and asterisk (*) to highlight in the spatial map.

For adjacent circles with limited space between them to display the background color, we use an economic implementation, colored outlines for these circles (\textbf{Figure~\ref{fig:visual}C}), inspired by watercolor effect \cite{pinna_1987, pinna_2001}. Watercolor effect describes the phenomenon in visual perception that surface color arises from thin boundaries and hence is applied here to perceive the background color in tight space. Overall, the figure-ground segmentation creates two isolated layers in visual perception to display the two variables while maintaining the relative spatial relationship serving as a reference between the two. In addition, other fundamental principles \cite{todorovic_2008}, such as proximity, similarity, continuity, and closure, incentivize the brain to group elements and dimensions in the visualization, guaranteeing an integrative perception of the complex multi-dimensional spatial map.

\begin{figure*}[t]
\begin{center}
\includegraphics[width=\textwidth]{Manuscript/figure/embedding.jpg}
\caption{\small \textbf{\texttt{escheR} improves multi-dimensional embedding visualizations}. The gene expression of \textit{POMGNT1} among peripheral blood mononuclear cells \cite{PBMC} under the UMAP representation. (\textbf{A}) The \texttt{schex} R/Bioconductor package uses color-coded convex hulls to annotate data-driven cell types, creating confusion when interpreting hexagons in overlapping hulls. (\textbf{B}) \texttt{escheR} plots hexagon-specific membership to avoid substantial overlapping of cluster membership. Due to the binning strategy, cautions in interpretation are needed due to possible cluster intermixing within each hexagon.} 
\label{fig:embedding} 
\end{center}
\vspace{-0.3in}
\end{figure*}

Here, we provide an open-source package called \texttt{escheR} (named after the graphic artist M.C. Escher) in the R programming language \cite{R}, leading to a simplified interface to navigate the implementation of the multi-dimensional visualization in 2D space. By adapting \texttt{ggplot2} standard, \texttt{escheR} can be seamlessly integrated with popular data objects, including \texttt{SingleCellExperiment} \cite{amezquita2020} and \texttt{SpatialExperiment} \cite{righelli2022}, and easily work with popular pipelines, such as \texttt{Seurat} \cite{hao_2021}, \texttt{Giotto} \cite{dries_2021} to name a few, and allow further theme customization with ease.

Next, we give two use cases to exemplify some utility of the proposed spatial visualization: (i) the spatially differential gene colocalization in the human dorsolateral prefrontal cortex using spatial transcriptomics data \cite{huukimyers_2023}; (ii) multi-dimensional UMAP highlighting differential gene expression in data-driven cell clusters \cite{freytag_2020}. 

\subsection{Multi-dimensional \textit{in situ} visualization}
In a recent study investigating the molecular organization of human dorsolateral prefrontal cortex \cite{huukimyers_2023}, two schizophrenia risk genes, membrane-bound ligand ephrin A5 (\textit{EFNA5}) and ephrin type-A receptor 5 (\textit{EPHA5}), were identified to colocalize via cell-cell communication analysis. In addition, data suggested Layer 6 was the most highly co-localized layer compared to other cortex layers. To visually examine the inference, we applied \texttt{escheR} to create a multi-dimensional \textit{in situ} spatial map that simultaneously exhibits the cortex layers (displayed with color-coded spot outlines) and the categorized colocalization status of genes \textit{EFNA5} and \textit{EPHA5} (displayed with color-coded spot fill). Compared to the traditional visualization where the cortex layers and the colocalization status are visualized in two side-by-side figures (\textbf{Figure \ref{fig:visual}A-B}), our proposed visualization (\textbf{Figure \ref{fig:visual}D}) enables directly mapping colocalization status to the spatial domain, simplifying the perception of two sources of information and allowing cognitive comparison across cortex layers. 

\subsection{Multi-dimensional embedding visualizations}

The application of the proposed framework is not limited to \textit{in situ} visualizations of spatially-resolved transcriptomics data. It is broadly applicable to data mapped to any 2-dimensional coordinate system to simultaneously display multiple variables. Such systems include euclidean space (including spatial coordinate as a special case) and data-driven embedding space, for example, UMAP and $t$-SNE. To demonstrate, we applied the proposed visualization to address the challenge of simultaneously displaying cluster membership and gene expression in a single-cell UMAP plot. To address the overplotting problem, previous work proposed to apply hexagonal binning strategy to display the gene expression \cite{freytag_2020}. Here, the color-coded convex hulls are used to annotate different clusters of cells (\textbf{Figure~\ref{fig:embedding}A}). However, the convex hulls create substantial overlapping areas, creating confusion when interpreting cluster memberships of hexagons in the overlapping areas. To improve the interpretability of the visualization, we replace the convex hulls with color-coded hexagons boundaries (\textbf{Figure~\ref{fig:embedding}B}) to avoid possible membership confusion. We note that our contribution to improving the visualization is easily implemented without any modification of \texttt{schex} as both are built upon the Grammar of Graphics \cite{wilkinson_2012} standard.


\section{Discussion}
Here, we propose an innovative multi-dimensional spatial visualization that simultaneously displays multiple variables in a 2D coordinate system. Specifically, our design leverages Gestalt principles from visual perception to create multiple visual dimensions in a spatial map by iterative layering aesthetics. Built upon \texttt{ggplot2}, we provide an open-source R package \texttt{escheR} that is seamlessly compatible with popular spatially-resolved transcriptomics and single-cell data analysis toolboxes. 

Adding a third dimension to 2D plots has been a long-standing challenge in visualization \cite{odonoghue_2010}. Our proposal addresses this fundamental challenge by introducing simple but effective design principles. These principles lead to visually easier-to-interpret graphics compared to 2D-color gradients and geometry annotations (\textbf{Figure~\ref{fig:embedding}}). Unlike computer-based interactive visualizations, the proposed visualization is free from any platform and technology restriction, creating an accessible and economical solution. In addition, the proposed visualization is easily scalable and hence can be applied to all types of spatially resolved data. 

Beyond the scope of biomedical research, the proposed visualization can be broadly translated to any visual analytic highlighting differentiation with respect to another measurement(s). To name a few, such visual analytics include examining differential tests,  explaining clustering, and visualizing subgroups. However, one of the most rewarding fields to apply the proposed visualization is the rapidly expanding field of biomedical multi-omics research \cite{hasin_2017}, where connecting different omics (data modalities) is the fundamental goal and hence greatly appreciating innovative multi-dimensional visualization. 

While the proposed design principle set-up a solid foundation, the proposed package \texttt{escheR} can be further optimized for image-based spatially resolved data \cite{caicedo_2017}. For example, to address possible overplotting of densely arranged single cells in tight space, strategies such as binning or hatching plots \cite{patrick2023spatial} can be incorporated. In addition, the flexibility to represent data points as polygons will be provided to allow visualization of cell morphology.
% Combining with a binning strategy, similar to \texttt{schex} \cite{freytag_2020} to avoid possible overplotting \rr{or hatching plot proposed in \texttt{lisaClust} \cite{patrick2023spatial}}, the proposed visualization can be also applied to visualized image-based spatially resolved data , in addition to aforementioned spot-based spatially resolved data (\textbf{Figure~\ref{fig:visual}}).  \rr{\texttt{escheR} could also be applied with kernel gene-weighted density plots from \texttt{Nebulosa} \cite{alquicirahernandez_2021} and other ggplot2-based visualizations.}
% \rr{Independently, \texttt{lisaClust} \cite{patrick2023spatial} visualize cell types within spatial domains by generating hatching plots for immunofluorescence and mass cytometry imaging data.} 

In summary, we propose a novel multi-dimensional visualization, implemented in an R package \texttt{escheR}, to address the simultaneous exhibition of multiple variables in 2D plots. The proposed visualization can be broadly applicable to the visual analytics of growingly complex biomedical data and beyond.






%% BIBLIOGRAPHY ==========================================================

% \printbibliography
\bibliography{refs}


\clearpage
%% EPILOGUE ================================================================
\section*{Abbreviations}

\begin{itemize}[nosep]
    \item \textbf{PCA}: principal component analysis
    \item \textbf{$t$-SNE}: $t$-distributed stochastic neighbor embedding
    \item \textbf{UMAP}: uniform manifold approximation and projection
    \item \textbf{scRNA-seq}: single-cell RNA-sequencing
    \item \textbf{snRNA-seq}: single-nucleus RNA-sequencing
    \item \textbf{SRT}: spatially-resolved transcriptomics 
    \item \textbf{EFNA5}: membrane-bound ligand ephrin A5
    \item \textbf{EPHA5}: ephrin type-A receptor 5
    \item \textbf{DLPFC}: dorsolateral prefrontal cortex
\end{itemize}

\section*{Author contributions}

\begin{itemize}[nosep]
    \item \textbf{Boyi Guo}:  Conceptualization, Methodology, Software, Validation, Formal analysis, Investigation, Data Curation, Writing, Visualization
    \item \textbf{Louise A. Huuki-Myers}: Conceptualization, Software
    \item \textbf{Melissa Grant-Peters}: Conceptualization
    \item \textbf{Leonardo Collado-Torres}: Conceptualization, Software
    \item \textbf{Stephanie C. Hicks}: Conceptualization, Resources, Writing - Review \& Editing, Visualization, Supervision, Project administration, Funding acquisition
\end{itemize}

\section*{Declarations}

\subsection*{Ethics approval and consent to participate}
Not applicable.

\subsection*{Competing interests}
The authors declare that they have no competing interests.

\subsection*{Availability of data and materials}
The spatial transcriptomics dataset was obtained from \texttt{spatialLIBD} (\href{http://research.libd.org/spatialLIBD}{research.libd.org/spatialLIBD}). The UMAP example follows the `using\_schex' vignette in the \texttt{schex} package (\href{https://www.bioconductor.org/packages/schex}{bioconductor.org/packages/schex}). The code that generates these figures is deposited at \href{https://github.com/boyiguo1/Manuscript_escheR}{github.com/boyiguo1/Manuscript\_escheR} (Zenodo DOI: \href{https://zenodo.org/record/7915970}{10.5281/zenodo.7915970}). The open source software package \texttt{escheR} available in the R programming language is freely available on GitHub (\href{https://github.com/boyiguo1/escheR}{github.com/boyiguo1/escheR}) and Bioconductor (\href{https://bioconductor.org/packages/escheR}{bioconductor.org/packages/escheR}).


\subsection*{Funding}
This project was supported by the National Institute of Mental Health [R01MH126393 to BG and SCH, U01MH122849 to LAHM and LCT]; the Chan Zuckerberg Initiative DAF, an advised fund of Silicon Valley Community Foundation [CZF2019-002443 to SCH]; the Lieber Institute for Brain Development to LAHM and LCT; and Aligning Science Across Parkinson’s [ASAP-000478, ASAP-000509 to MGP] through the Michael J. Fox Foundation for Parkinson’s Research.  All funding bodies had no role in the design of the study and collection, analysis, and interpretation of data and in writing the manuscript.

\subsection*{Acknowledgments}
We would like to acknowledge Nicholas J. Eagles, Kristen R. Maynard, Mina Ryten, Leon Di Stefano, and Lukas M. Weber (appearing in alphabetic order of last name) for their helpful comments, feedback and suggestions on \texttt{escheR} functionality. Nicholas J. Eagles and Kristen R. Maynard are employed by the Lieber Institute for Brain Development; Mina Ryten is employed by University College London; Leon Di Stefano and Lukas M. Weber are from the Johns Hopkins Bloomberg School of Public Health, Department of Biostatistics. 


\subsection*{Author’s information}

\begin{itemize}[nosep]
    \item Boyi Guo (ORCiD: \href{https://orcid.org/0000-0003-2950-2349}{0000-0003-2950-2349})
    \item Erik D. Nelson (ORCiD: \href{https://orcid.org/0000-0001-8477-0982}{0000-0001-8477-0982})
    \item Louise A. Huuki-Myers (ORCiD: \href{https://orcid.org/0000-0001-5148-3602}{0000-0001-5148-3602})
    \item Melissa Grant-Peters (ORCiD: \href{https://orcid.org/0000-0003-0585-0971}{0000-0003-0585-0971})
    \item Leonardo Collado-Torres (ORCiD: \href{https://orcid.org/0000-0003-2140-308X}{0000-0003-2140-308X})
    \item Stephanie C. Hicks (ORCiD: \href{https://orcid.org/0000-0002-7858-0231}{0000-0002-7858-0231})
\end{itemize} 

\subsection*{Conflict of Interest} 
None declared.
% \onecolumn

% {\huge Supplementary Materials}

% \hrule

% \vspace*{0.5cm}

% \begin{center}

% {\Large \texttt{escheR}: Unified multi-dimensional visualizations with Gestalt principles}

% \vspace*{0.75cm}

% {\large Boyi Guo, Louise A. Huuki-Myers, Melissa Grant-Peters, Leonardo Collado-Torres, Stephanie C.\ Hicks$^*$}

% \vspace*{0.3cm}

% {\small $^*$Correspondence to \url{shicks19@jhu.edu}}

% \end{center}

% \renewcommand{\figurename}{Supplementary Figure}
% \renewcommand{\tablename}{Supplementary Table}
% \setcounter{figure}{0}
% \setcounter{table}{0}
% \setcounter{section}{0}
% \setcounter{page}{1}
% \makeatletter
% \renewcommand{\thefigure}{S\@arabic\c@figure}
% \renewcommand{\thetable}{S\@arabic\c@table}
% \renewcommand{\thesection}{Supplemental Note S\@arabic\c@section}
% \makeatother

% \vspace*{1cm}

% {\bf \large Contents}

% % \begin{enumerate}
%     \item \textbf{Supplemental Figures~S1.}
% \end{enumerate}

% \clearpage 

% \noindent {\LARGE Supplemental Figures}





\end{document}


